### DISCUSSION TITLE
`vcpkg new` CLI command for new project initiation

### DISCUSSION THREAD

**Discussion by patrolez** [microsoft/vcpkg#17903] [Ideas]
CLI as Command-line interface, not ~Common Language Infrastructure~.

****

**Is your feature request related to a problem? Please describe.**
I'm always frustrated when ... I have to every time create minimal "Hello World!" grade project with just CMakeLists.txt with basic filesystem organization and adjustments to make it working with `vcpkg`.

While using `dotnet` CLI I am glad to have `dotnet new` opportunity. [(link)](https://docs.microsoft.com/en-au/dotnet/core/tools/dotnet-new)
While using `cargo` CLI for Rust language experience I can use `cargo new`. [(link)](https://doc.rust-lang.org/cargo/guide/creating-a-new-project.html#creating-a-new-package)
While using `npm` CLI I can use `npm init`. [(link)](https://docs.npmjs.com/cli/init)
While using `angular` CLI I can use `ng new`. [(link)](https://angular.io/cli/new)
I am not using `conan`, but seems like there exists `conan new`. ([link](https://docs.conan.io/en/latest/reference/commands/creator/new.html))

**Proposed solution**
`vcpkg new` based on `dotnet new` architectural principles (including template system).

**Describe alternatives you've considered**
Nothing exists regarding C++. I can only rely on some repositories called like "project template" or "minimal project template": https://github.com/search?o=desc&q=cmake+project+template&s=stars&type=Repositories

https://vcpkg.readthedocs.io/en/latest/specifications/manifests/ could contain prepacked archives to bootstrap a project immediately after download-unpack-rename.

https://vcpkg.readthedocs.io/en/latest/examples/installing-and-using-packages/#cmake could contain prepacked archives to bootstrap a project immediately after download-unpack-rename.

**Additional context**
* Personally I think it would work better with `manifest` way of working regarding `vcpkg`.
* If `project template` system would be a huge thing, so I think that development could be divided for some basic system and more generic one.
* Parametrized `project templates` could be nice thing for example to providing C++ language version.
   * parametrization also could take into account MSBuild, CMake and provide some way to implement another integrations by contributions.
   * parametrization also might take into account the most popular Version control systems and/or `vcpkg as submodule` initiated project

**Comment by StarGate-One** [comment]
Have you explored **`vcpkg create <pkg> <url> [archivename] create a new package`**?

**Comment by patrolez** [comment]
> Have you explored **`vcpkg create <pkg> <url> [archivename] create a new package`**?

Neither <https://vcpkg.readthedocs.io/en/latest/specifications/manifests/>, nor <https://vcpkg.readthedocs.io/en/latest/examples/installing-and-using-packages/#cmake> indicates that it can be done this way.

I have tried to use that, but only new directory appeared in `vcpkg/ports/` which is not usable in terms of what I have had on my mind by introducing this topic.

`vcpkg create` it made for different purposes: https://vcpkg.readthedocs.io/en/latest/examples/packaging-zipfiles/

As I understood docs linked above, this command allows you to use some bunch of code as a vcpkg library which is not already maintained as a library available under `vcpkg search`.

**Comment by JackBoosY** [comment]
Related: #9430

**Comment by patrolez** [comment]
Hello. As @strega-nil was assigned I discovered a repo https://github.com/strega-nil/vcpkg-example which is the thing I was thinking about to be "generated" on anyone's local machine after using command like:
`vcpkg new --feature-flags=binarycaching,manifests --vcpkg-resolve-deps bin-hello-world-fmt-git-example --ci-githubworkflow --cmake vcpkg-example`.

Where I assumed that:
❶ `--feature-flags=binarycaching,manifests --vcpkg-resolve-deps` would be some options passed and interpreted by vcpkg;
❷ `bin-hello-world-fmt-git-example` would be a name for a project template followed by parameters passed to a project template generator (customizable by project template maintainer, in this case flavor with `git`, `fmt` and `binary executable` target);
❸ `vcpkg-example` would be a name for final cmake@vcpkg@git@C++ project and the first executable target linked to src/main.cpp OR src/main.cxx under directory with the same name by default (or possibly changed by some option passed in ❶ CLI section).

**Comment by strega-nil** [comment]
@patrolez yeah, I like this! Would you be interested in working on it?
